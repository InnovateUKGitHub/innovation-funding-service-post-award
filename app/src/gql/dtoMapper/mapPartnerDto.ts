import { SalesforceProjectRole } from "@framework/constants/salesforceProjectRole";
import type { PartnerDtoGql } from "@framework/dtos";
import { getClaimStatus } from "@framework/mappers/claimStatus";
import { getPartnerStatus } from "@framework/mappers/partnerStatus";
import { calcPercentage } from "@framework/util";

type PartnerNode = Readonly<
  Partial<{
    Id: string;
    Acc_AccountId__c: GQL.Value<string>;
    Acc_AccountId__r: {
      Name: GQL.Value<string>;
      Id?: string | null;
    } | null;
    Acc_AuditReportFrequency__c: GQL.Value<string>;
    Acc_Award_Rate__c: GQL.Value<number>;
    Acc_Cap_Limit__c: GQL.Value<number>;
    Acc_ForecastLastModifiedDate__c: GQL.Value<string>;
    Acc_NewForecastNeeded__c: GQL.Value<boolean>;
    Acc_NonfundedParticipant__c: GQL.Value<boolean>;
    Acc_OpenClaimStatus__c: GQL.Value<string>;
    Acc_OrganisationType__c: GQL.Value<string>;
    Acc_OverheadRate__c: GQL.Value<number>;
    Acc_ParticipantStatus__c: GQL.ValueAndLabel<string>;
    Acc_ParticipantType__c: GQL.Value<string>;
    Acc_Postcode__c: GQL.Value<string>;
    Acc_ProjectRole__c: GQL.Value<string>;
    Acc_RemainingParticipantGrant__c: GQL.Value<number>;
    Acc_TotalApprovedCosts__c: GQL.Value<number>;
    Acc_TotalCostsSubmitted__c: GQL.Value<number>;
    Acc_TotalFutureForecastsForParticipant__c: GQL.Value<number>;
    Acc_TotalGrantApproved__c: GQL.Value<number>;
    Acc_TotalParticipantCosts__c: GQL.Value<number>;
    Acc_TotalPrepayment__c: GQL.Value<number>;
    Acc_TrackingClaims__c: GQL.Value<string>;
  }>
> | null;

type PartnerDtoMapping = Pick<
  PartnerDtoGql,
  | "id"
  | "accountId"
  | "auditReportFrequencyName"
  | "awardRate"
  | "capLimit"
  | "claimStatus"
  | "competitionName"
  | "forecastLastModifiedDate"
  | "forecastsAndCosts"
  | "isLead"
  | "isNonFunded"
  | "isWithdrawn"
  | "name"
  | "newForecastNeeded"
  | "organisationType"
  | "overheadRate"
  | "partnerStatus"
  | "partnerStatusLabel"
  | "percentageParticipantCostsClaimed"
  | "percentageParticipantCostsSubmitted"
  | "postcode"
  | "projectId"
  | "remainingParticipantGrant"
  | "roles"
  | "totalCostsSubmitted"
  | "totalGrantApproved"
  | "totalParticipantCostsClaimed"
  | "totalParticipantGrant"
  | "totalPrepayment"
  | "type"
>;

const mapper: GQL.DtoMapper<PartnerDtoMapping, PartnerNode, { roles?: SfRoles; competitionName?: string }> = {
  id(node) {
    return (node?.Id ?? "") as PartnerId;
  },
  accountId(node) {
    return node?.Acc_AccountId__c?.value ?? "";
  },
  auditReportFrequencyName(node) {
    return node?.Acc_AuditReportFrequency__c?.value ?? "";
  },
  awardRate(node) {
    return node?.Acc_Award_Rate__c?.value ?? null;
  },
  capLimit(node) {
    return node?.Acc_Cap_Limit__c?.value ?? 0;
  },
  claimStatus(node) {
    return getClaimStatus(node?.Acc_TrackingClaims__c?.value ?? "");
  },
  competitionName(node, additionalData) {
    return additionalData.competitionName;
  },
  forecastLastModifiedDate(node) {
    const dateValue = node?.Acc_ForecastLastModifiedDate__c?.value;
    if (dateValue === null) return null;
    if (dateValue === undefined) return null;
    return new Date(dateValue);
  },
  forecastsAndCosts(node) {
    return (
      (node?.Acc_TotalFutureForecastsForParticipant__c?.value ?? 0) + (node?.Acc_TotalCostsSubmitted__c?.value ?? 0)
    );
  },
  isLead(node) {
    return node?.Acc_ProjectRole__c?.value === SalesforceProjectRole.ProjectLead;
  },
  isNonFunded(node) {
    return node?.Acc_NonfundedParticipant__c?.value ?? false;
  },
  isWithdrawn(node) {
    return ["Voluntary Withdrawal", "Involuntary Withdrawal", "Migrated - Withdrawn"].includes(
      node?.Acc_ParticipantStatus__c?.value ?? "",
    );
  },
  name(node) {
    return node?.Acc_AccountId__r?.Name?.value ?? "Unknown name";
  },
  newForecastNeeded(node) {
    return node?.Acc_NewForecastNeeded__c?.value ?? false;
  },
  organisationType(node) {
    return node?.Acc_OrganisationType__c?.value ?? "unknown";
  },
  overheadRate(node) {
    return node?.Acc_OverheadRate__c?.value ?? null;
  },
  partnerStatus(node) {
    return getPartnerStatus(node?.Acc_ParticipantStatus__c?.value ?? "unknown");
  },
  partnerStatusLabel(node) {
    return node?.Acc_ParticipantStatus__c?.label ?? "unknown";
  },
  percentageParticipantCostsClaimed(node) {
    return calcPercentage(node?.Acc_TotalParticipantCosts__c?.value ?? 0, node?.Acc_TotalApprovedCosts__c?.value ?? 0);
  },
  percentageParticipantCostsSubmitted(node) {
    return calcPercentage(node?.Acc_TotalParticipantCosts__c?.value ?? 0, node?.Acc_TotalCostsSubmitted__c?.value ?? 0);
  },
  postcode(node) {
    return node?.Acc_Postcode__c?.value ?? null;
  },
  projectId(node) {
    return (node?.Acc_AccountId__r?.Id ?? "") as ProjectId;
  },
  remainingParticipantGrant(node) {
    return node?.Acc_RemainingParticipantGrant__c?.value ?? 0;
  },
  roles(node, additionalData: { roles?: SfRoles }) {
    return additionalData?.roles ?? { isPm: false, isMo: false, isFc: false };
  },
  totalCostsSubmitted(node) {
    return node?.Acc_TotalCostsSubmitted__c?.value ?? null;
  },
  totalGrantApproved(node) {
    return node?.Acc_TotalGrantApproved__c?.value ?? null;
  },
  totalParticipantCostsClaimed(node) {
    return node?.Acc_TotalApprovedCosts__c?.value ?? null;
  },
  totalParticipantGrant(node) {
    return node?.Acc_TotalParticipantCosts__c?.value ?? null;
  },
  totalPrepayment(node) {
    return node?.Acc_TotalPrepayment__c?.value ?? null;
  },
  type(node) {
    return node?.Acc_ParticipantType__c?.value ?? "unknown";
  },
};

type PartnerAdditionalData<TPickList extends string> = AdditionalDataType<
  TPickList,
  [["roles", "roles", SfRoles], ["competitionName", "competitionName", string]]
>;

/**
 * Maps a specified Partner Node from a GQL query to a slice of
 * the PartnerDto to ensure consistency and compatibility in the application
 */
export function mapToPartnerDto<TNode extends PartnerNode, TPickList extends keyof PartnerDtoMapping>(
  projectNode: TNode,
  pickList: TPickList[],
  additionalData: PartnerAdditionalData<TPickList>,
): Pick<PartnerDtoMapping, TPickList> {
  return pickList.reduce((dto, field) => {
    dto[field] = mapper[field](projectNode, additionalData);
    return dto;
  }, {} as Pick<PartnerDtoMapping, TPickList>);
}

const defaultRole = { isPm: false, isMo: false, isFc: false };

/**
 * Maps Partner Edges to array of Partner DTOs.
 *
 * PartnerRoles object must be supplied as separate argument
 */
export function mapToPartnerDtoArray<
  TNode extends ReadonlyArray<{ node: PartnerNode } | null> | null,
  TPickList extends keyof PartnerDtoMapping,
>(
  partnerEdges: TNode,
  pickList: TPickList[],
  additionalData: AdditionalDataType<
    TPickList,
    [["roles", "partnerRoles", SfPartnerRoles[]], ["competitionName", "competitionName", string]]
  >,
): Pick<PartnerDtoMapping, TPickList>[] {
  return (
    partnerEdges?.map(node => {
      if (additionalData && "partnerRoles" in additionalData) {
        const { partnerRoles, ...nextAdditionalData } = additionalData;
        const roles: SfRoles =
          (partnerRoles as SfPartnerRoles[]).find(x => x?.partnerId === node?.node?.Acc_AccountId__c?.value) ??
          defaultRole;
        return mapToPartnerDto(node?.node ?? null, pickList, {
          ...nextAdditionalData,
          roles,
        } as unknown as PartnerAdditionalData<TPickList>);
      }
      return mapToPartnerDto(
        node?.node ?? null,
        pickList,
        additionalData as unknown as PartnerAdditionalData<TPickList>,
      );
    }) ?? []
  );
}
