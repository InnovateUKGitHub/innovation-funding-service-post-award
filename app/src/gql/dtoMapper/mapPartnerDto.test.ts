import { mapToPartnerDtoArray } from "./mapPartnerDto";

describe("mapPartnerDtoArray", () => {
  const edges = [
    {
      node: {
        Id: "0",
        Acc_ProjectRole__c: { value: "Lead" },
        Acc_ParticipantStatus__c: { value: "Active" },
        Acc_AccountId__r: {
          Name: { value: "Swindon University" },
          Id: "123",
        },
        Acc_AccountId__c: { value: "123" },
        Acc_TotalParticipantCosts__c: { value: 10000 },
        Acc_TotalApprovedCosts__c: { value: 20000 },
        Acc_NewForecastNeeded__c: { value: true },
        Acc_TrackingClaims__c: { value: "Claims Overdue" },
        Acc_OverheadRate__c: { value: 100 },
        Acc_OrganisationType__c: { value: "Academic" },
        Acc_ForecastLastModifiedDate__c: { value: "2022-01-24" },
        Acc_TotalFutureForecastsForParticipant__c: { value: 10000 },
        Acc_TotalCostsSubmitted__c: { value: 5000 },
        Acc_TotalGrantApproved__c: { value: 100000 },
        Acc_Award_Rate__c: { value: 10 },
        Acc_RemainingParticipantGrant__c: { value: 3000 },
        Acc_TotalPrepayment__c: { value: 4000 },
        Acc_Cap_Limit__c: { value: 100 },
        Acc_AuditReportFrequency__c: { value: "monthly" },
        Acc_OpenClaimStatus__c: { value: "Draft" },
      },
    },
    {
      node: {
        Id: "1",
        Acc_ProjectRole__c: { value: "Regular" },
        Acc_ParticipantStatus__c: { value: "Voluntary Withdrawal" },
        Acc_AccountId__r: {
          Name: { value: "Kettering Nursing College" },
          Id: "124",
        },
        Acc_AccountId__c: { value: "124" },
        Acc_TotalParticipantCosts__c: { value: 10000 },
        Acc_TotalApprovedCosts__c: { value: 20000 },
        Acc_NewForecastNeeded__c: { value: true },
        Acc_TrackingClaims__c: { value: "Claims Overdue" },
        Acc_OverheadRate__c: { value: 100 },
        Acc_OrganisationType__c: { value: "Academic" },
        Acc_ForecastLastModifiedDate__c: { value: "2022-01-24" },
        Acc_TotalFutureForecastsForParticipant__c: { value: 10000 },
        Acc_TotalCostsSubmitted__c: { value: 5000 },
        Acc_TotalGrantApproved__c: { value: 100000 },
        Acc_Award_Rate__c: { value: 10 },
        Acc_RemainingParticipantGrant__c: { value: 3000 },
        Acc_TotalPrepayment__c: { value: 4000 },
        Acc_Cap_Limit__c: { value: 100 },
        Acc_AuditReportFrequency__c: { value: "monthly" },
        Acc_OpenClaimStatus__c: { value: "New" },
      },
    },
  ];
  const partnerRoles = [
    { isFc: true, isPm: true, isMo: false, partnerId: "123" },
    { isFc: true, isPm: false, isMo: false, partnerId: "124" },
  ];

  it("should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records", () => {
    expect(
      mapToPartnerDtoArray(
        edges,
        [
          "id",
          "accountId",
          "auditReportFrequencyName",
          "awardRate",
          "capLimit",
          "claimStatus",
          "competitionName",
          "forecastLastModifiedDate",
          "forecastsAndCosts",
          "isLead",
          "isWithdrawn",
          "name",
          "newForecastNeeded",
          "organisationType",
          "overheadRate",
          "partnerStatus",
          "percentageParticipantCostsClaimed",
          "percentageParticipantCostsSubmitted",
          "remainingParticipantGrant",
          "roles",
          "totalCostsSubmitted",
          "totalGrantApproved",
          "totalParticipantCostsClaimed",
          "totalParticipantGrant",
          "totalPrepayment",
        ],
        { partnerRoles, competitionName: "bingo" },
      ),
    ).toMatchSnapshot();
  });

  it("should map the gql data only including fields on the picklist", () => {
    expect(mapToPartnerDtoArray(edges, ["id", "roles", "isLead"], { partnerRoles })).toMatchSnapshot();
  });
});
