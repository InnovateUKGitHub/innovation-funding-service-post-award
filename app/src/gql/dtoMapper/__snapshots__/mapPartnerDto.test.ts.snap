// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapPartnerDtoArray should map the gql data only including fields on the picklist 1`] = `
Array [
  Object {
    "id": "0",
    "isLead": true,
    "roles": Object {
      "isFc": true,
      "isMo": false,
      "isPm": true,
      "partnerId": "123",
    },
  },
  Object {
    "id": "1",
    "isLead": false,
    "roles": Object {
      "isFc": true,
      "isMo": false,
      "isPm": false,
      "partnerId": "124",
    },
  },
]
`;

exports[`mapPartnerDtoArray should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records 1`] = `
Array [
  Object {
    "accountId": "123",
    "auditReportFrequencyName": "monthly",
    "awardRate": 10,
    "capLimit": 100,
    "claimStatus": 3,
    "competitionName": "bingo",
    "forecastLastModifiedDate": 2022-01-24T00:00:00.000Z,
    "forecastsAndCosts": 15000,
    "id": "0",
    "isLead": true,
    "isWithdrawn": false,
    "name": "Swindon University",
    "newForecastNeeded": true,
    "organisationType": "Academic",
    "overheadRate": 100,
    "partnerStatus": 1,
    "percentageParticipantCostsClaimed": 200,
    "percentageParticipantCostsSubmitted": 50,
    "remainingParticipantGrant": 3000,
    "roles": Object {
      "isFc": true,
      "isMo": false,
      "isPm": true,
      "partnerId": "123",
    },
    "totalCostsSubmitted": 5000,
    "totalGrantApproved": 100000,
    "totalParticipantCostsClaimed": 20000,
    "totalParticipantGrant": 10000,
    "totalPrepayment": 4000,
  },
  Object {
    "accountId": "124",
    "auditReportFrequencyName": "monthly",
    "awardRate": 10,
    "capLimit": 100,
    "claimStatus": 3,
    "competitionName": "bingo",
    "forecastLastModifiedDate": 2022-01-24T00:00:00.000Z,
    "forecastsAndCosts": 15000,
    "id": "1",
    "isLead": false,
    "isWithdrawn": true,
    "name": "Kettering Nursing College",
    "newForecastNeeded": true,
    "organisationType": "Academic",
    "overheadRate": 100,
    "partnerStatus": 4,
    "percentageParticipantCostsClaimed": 200,
    "percentageParticipantCostsSubmitted": 50,
    "remainingParticipantGrant": 3000,
    "roles": Object {
      "isFc": true,
      "isMo": false,
      "isPm": false,
      "partnerId": "124",
    },
    "totalCostsSubmitted": 5000,
    "totalGrantApproved": 100000,
    "totalParticipantCostsClaimed": 20000,
    "totalParticipantGrant": 10000,
    "totalPrepayment": 4000,
  },
]
`;
